lib/miq_automation_engine/engine/miq_ae_engine.rb

```diff
@@ -51,7 +51,13 @@ def self.deliver_queue(args, options = {})
   end
 
   def self.deliver(*args)
-    @deliver_mutex.synchronize { deliver_block(*args) }
+    # This was added because in RAILS 5 PUMA sends multiple requests
+    # to the same processa and our DRb server implementation for Automate
+    # Methods is not thread safe. The permit_concurrent_loads  allows the
+    # DRb requests which run in different threads to load symbols.
```

Comment:
> So these are not really auto-load, if we had loaded them all upfront we wouldn't have seen this issue, these constants come about when some of the constants are missing and Rails tries to load them.

Right, @mkanoor.  "Loading constants" is clearer than "loading symbols". 

The key point is if you don't eager load all constants to be used in your automate "methods/dialogs/things" and rely on Rails to autoload those missing constants, you could encounter a race condition with multiple threads, where the holders of two locks cause a deadlock:

* `ActiveSupport::Dependencies.interlock`
* `@deliver_mutex`

This permit_concurrent_loads allows constant loading by yielding 'share' locks to avoid deadlocks:
```
+      # Temporarily give up all held Share locks while executing the
+      # supplied block, allowing any +compatible+ exclusive lock request
+      # to proceed.
```

From [here](https://github.com/rails/rails/pull/23532/files#diff-d252ad9aa6114314925757c521598dbeR134)